### PR TITLE
openvscode-server: init at 1.62.0

### LINKED ISSUE #140344: package request: openvscode-server
**Project description**

It's like `code-server` but closer to upstream, as I understand it.

Maybe this can be packaged similarly to how `code-server` is packaged?

shamelessly ccing past code-server contributors:  @offlinehacker  @dguenther (hi!) and maybe @ghuntley would be interested, or maybe has a nix derivation already?

**Metadata**

* homepage URL: https://github.com/gitpod-io/openvscode-server/
* source URL: https://github.com/gitpod-io/openvscode-server/
* license: **mit**
* platforms: **linux**

### PR DESCRIPTION
**Pull request by dguenther**
###### Motivation for this change

The derivation is based on the work done for [code-server](https://github.com/NixOS/nixpkgs/blob/17f230461a1010cdb0d0bb92594546429c69de47/pkgs/servers/code-server/default.nix), but is structured as a fork of upstream VS Code instead of a series of nested NPM packages.

ccing @colemickens as the creator of #140344

Closes #140344

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dguenther** [comment]
Putting this in draft for a bit, I bumped the version to 1.62 and need to fix some build steps.

File: pkgs/servers/openvscode-server/default.nix [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    '';
+
+    installPhase = ''
+      echo yarnCache
+    '';
+
+    outputHashMode = "recursive";
+    outputHashAlgo = "sha256";
+
+    # to get hash values use nix-build -A vscode-webserver.prefetchYarnCache
```
Comment (ghuntley) [inline pkgs/servers/openvscode-server/default.nix:47]:
```suggestion
    # to get hash values use nix-build -A openvscode-server.prefetchYarnCache
```

File: pkgs/servers/openvscode-server/default.nix [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+  passthru = {
+    prefetchYarnCache = lib.overrideDerivation yarnCache (d: {
+      outputHash = lib.fakeSha256;
+    });
+  };
+
+  meta = with lib; {
+    description = "Run VS Code on a remote server";
+    longDescription = ''
+      vscode-webserver is VS Code running on a remote server, accessible through the
```
Comment (ghuntley) [inline pkgs/servers/openvscode-server/default.nix:179]:
```suggestion
      openvscode-server is VS Code running on a remote server, accessible through the
```

File: pkgs/servers/openvscode-server/default.nix [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+      browser.
+    '';
+    homepage = "https://github.com/gitpod-io/openvscode-server";
+    license = licenses.mit;
+    maintainers = with maintainers; [ offline ghuntley ];
```
Comment (ghuntley) [inline pkgs/servers/openvscode-server/default.nix:184]: ✅ Happy to help with maintenance, thanks for fixing the solution to esbuild post install scripts.

**Review by ghuntley — APPROVED** [review]

File: pkgs/servers/openvscode-server/default.nix (hunk 3 above)
Comment (dguenther) [reply pkgs/servers/openvscode-server/default.nix]: 👍 @offlinehacker I removed you as maintainer from this new package so as not to force the effort of maintenance on you -- happy to add you back if you'd like!

File: pkgs/servers/openvscode-server/default.nix [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+      yarn --offline config set ${option}
+    '') defaultYarnOpts}
+
+    # set offline mirror to yarn cache we created in previous steps
+    yarn --offline config set yarn-offline-mirror "${yarnCache}"
+
+    # skip unnecessary electron download
+    export ELECTRON_SKIP_BINARY_DOWNLOAD=1
+  '' + lib.optionalString stdenv.isLinux ''
+
```
[the remark this answers is not in the document]
Comment (dguenther) [reply pkgs/servers/openvscode-server/default.nix]: I think this comment was to remove whitespace? Should be updated now.

File: pkgs/servers/openvscode-server/default.nix [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    cp -R -T ../vscode-reh-web-${vsBuildTarget} "$out/libexec"
+
+    makeWrapper "${nodejs}/bin/node" "$out/bin/openvscode-server" \
+      --add-flags "$out/libexec/out/vs/server/main.js"
+  '';
+
+  passthru = {
+    prefetchYarnCache = lib.overrideDerivation yarnCache (d: {
+      outputHash = lib.fakeSha256;
```
[the remark this answers is not in the document]
Comment (dguenther) [reply pkgs/servers/openvscode-server/default.nix]: I deleted this section, from what I can tell it's not a common pattern in other packages.

**Comment by dguenther** [comment]
I simplified the build process a bit and removed some unused dependencies, in addition to fixing @parcel/watcher on darwin -- this should be ready for a re-review.

**Comment by colemickens** [comment]
@dguenther I'm sitting through a massive re-build, anything I should know for testing this? I was using this with code-server: https://github.com/cole-mickens/nixcfg/blob/43157e176fbf5914f31f7343cc29b9c24d7e4688/modules/code-server.nix (found somewhere online)

**Comment by dguenther** [comment]
I haven't yet investigated the config options and different folders it creates for settings/extensions, but this might be a good starting point for config options: https://github.com/gitpod-io/openvscode-server/blob/main/src/vs/server/serverEnvironmentService.ts

The main path I've followed for testing has been to launch the server directly, load the URL it outputs in a browser, try opening the terminal (tests that native code was built properly), then install the Nix extension and try opening the `nixpkgs` git repository as a folder to test file watching. So if you have any ideas for additional testing beyond that, that'd be great!

**Comment by colemickens** [comment]
Interesting, it seems like the auth is arguably both better and worse for this build than code-server. Better in that it feels like it actually enforces it a bit, and worse in that it's harder to just throw behind an auth'd proxy.

**Comment by colemickens** [comment]
Aha, seems like the instructions for usage is to run `server.sh`, which lives in here with other web-server-launcher-goodies: https://github.com/gitpod-io/openvscode-server/tree/main/resources/server

**Comment by colemickens** [comment]
I'll probably merge this soon, might be interesting to see if there's a number of us using this, maybe open a Matrix room and try to consolidate energy around one fork or the other.

**Comment by dguenther** [comment]
> Interesting, it seems like the auth is arguably both better and worse for this build than code-server. Better in that it feels like it actually enforces it a bit, and worse in that it's harder to just throw behind an auth'd proxy.

Is that regarding the connection token? There were some notes about that token being new to 1.62, so that may end up in code-server eventually as well in some form, if they decide to use the upstream server: https://github.com/gitpod-io/openvscode-server#a-note-about-security-tokens

> Aha, seems like the instructions for usage is to run `server.sh`, which lives in here with other web-server-launcher-goodies: https://github.com/gitpod-io/openvscode-server/tree/main/resources/server

Yep, this file specifically is what gets output and is supposed to be executed: https://github.com/gitpod-io/openvscode-server/blob/main/resources/server/bin/server.sh

I'll see if I can update the derivation to run this script instead, that should minimize the chance of breakage in the future.

**Comment by dguenther** [comment]
All right, updated it to run the `server.sh` script instead 👍

**Comment by colemickens** [comment]
While I'm using this successfully and it roughly LGTM, and given that @ghuntley signed off on an earlier review makes me feel comfortable merging this.

**Pull request merged**